File: sdk-ts/src/receipt.ts
/**
 * `Receipt` — the first of the SDK's two user-facing nouns.
 *
 * A `Receipt` is dregg's canonical proof-of-execution artifact: the committed
 * turn's hashes, the pre/post state roots, the agent cell, the federation
 * binding, and the receipt-chain link. Every authorized submission
 * (`AuthorizedTurn.submit()`) returns one, and the event stream
 * (`NodeEvents.subscribe`) yields the same noun — observation and action
 * speak one language.
 *
 * The proof is **lazily attached**: a `Receipt` is born proofless (the commit
 * decision is the executor's; the STARK is additive attestation) and a
 * [`TurnProof`] can be attached when the node's async prove pool produces one
 * (`fetchProof` pulls `/api/turn/{hash}/proof`).
 *
 * The second noun is `AttestedHistory` — the light-client whole-history
 * verdict. It is produced by verifying ONE succinct aggregate and is not yet
 * surfaced on the TS SDK (the verifier is Rust-side).
 */

import { bytesEqual, hexDecodeExact, hexEncode } from "./internal/bytes";

/** The wire-level canonical receipt as the node serializes it (serde JSON). */
export interface TurnReceiptJson {
  turn_hash: number[];
  forest_hash: number[];
  pre_state_hash: number[];
  post_state_hash: number[];
  timestamp: number;
  effects_hash: number[];
  computrons_used: number;
  action_count: number;
  previous_receipt_hash: number[] | null;
  agent: number[];
  federation_id?: number[];
  emitted_events?: unknown[];
  finality?: string;
  was_encrypted?: boolean;
  was_burn?: boolean;
  [extra: string]: unknown;
}

/**
 * The composed full-turn STARK attached to a receipt — opaque bytes bound to
 * the turn hash they attest (replay protection: a proof only attests the
 * turn it names). Verification happens Rust-side (`verify_full_turn`); the
 * TS SDK carries and transports the artifact.
 */
export class TurnProof {
  /** The turn hash this proof is bound to (32 bytes). */
  readonly turnHash: Uint8Array;
  /** The wire-serialized proof bytes. */
  readonly bytes: Uint8Array;

  constructor(turnHash: Uint8Array, bytes: Uint8Array) {
    if (turnHash.length !== 32) throw new Error("TurnProof: turnHash must be 32 bytes");
    this.turnHash = Uint8Array.from(turnHash);
    this.bytes = bytes;
  }

  get turnHashHex(): string {
    return hexEncode(this.turnHash);
  }
}

/** Thrown when an attachment names a different turn than the receipt. */
export class WrongTurnProofError extends Error {
  constructor(expectedHex: string, gotHex: string) {
    super(`proof is bound to turn ${gotHex}, receipt is turn ${expectedHex}`);
    this.name = "WrongTurnProofError";
  }
}

/** The normalized receipt fields every construction path can fill. */
export interface ReceiptFields {
  /** Hex turn hash (always present). */
  turnHash: string;
  /** Hex receipt hash, when the server reported it. */
  receiptHash?: string;
  /** Hex agent cell id. */
  agent?: string;
  preStateHash?: string;
  postStateHash?: string;
  timestamp?: number;
  computronsUsed?: number;
  actionCount?: number;
  previousReceiptHash?: string;
  finality?: string;
  wasEncrypted?: boolean;
  wasBurn?: boolean;
  /** Position in the node's receipt chain (SSE resume cursor). */
  chainIndex?: number;
  /** Whether the node reported an attached attestation at serve time. */
  hasProofHint?: boolean;
  /** The full canonical wire receipt, when available (SSE path). */
  raw?: TurnReceiptJson;
}

const num32ToHex = (a: number[] | null | undefined): string | undefined =>
  Array.isArray(a) ? hexEncode(Uint8Array.from(a)) : undefined;

/**
 * **The receipt noun.** A committed turn's proof-of-execution, with the
 * composed STARK proof lazily attached.
 */
export class Receipt {
  readonly turnHash: string;
  readonly receiptHash?: string;
  readonly agent?: string;
  readonly preStateHash?: string;
  readonly postStateHash?: string;
  readonly timestamp?: number;
  readonly computronsUsed?: number;
  readonly actionCount?: number;
  readonly previousReceiptHash?: string;
  readonly finality?: string;
  readonly wasEncrypted?: boolean;
  readonly wasBurn?: boolean;
  readonly chainIndex?: number;
  readonly hasProofHint?: boolean;
  readonly raw?: TurnReceiptJson;

  private attached: TurnProof | undefined;

  constructor(fields: ReceiptFields) {
    this.turnHash = fields.turnHash.toLowerCase();
    this.receiptHash = fields.receiptHash?.toLowerCase();
    this.agent = fields.agent?.toLowerCase();
    this.preStateHash = fields.preStateHash;
    this.postStateHash = fields.postStateHash;
    this.timestamp = fields.timestamp;
    this.computronsUsed = fields.computronsUsed;
    this.actionCount = fields.actionCount;
    this.previousReceiptHash = fields.previousReceiptHash;
    this.finality = fields.finality;
    this.wasEncrypted = fields.wasEncrypted;
    this.wasBurn = fields.wasBurn;
    this.chainIndex = fields.chainIndex;
    this.hasProofHint = fields.hasProofHint;
    this.raw = fields.raw;
  }

  /** Build from the canonical wire receipt (the SSE `receipt` field). */
  static fromTurnReceipt(r: TurnReceiptJson, extra?: Partial<ReceiptFields>): Receipt {
    return new Receipt({
      turnHash: num32ToHex(r.turn_hash) ?? "",
      agent: num32ToHex(r.agent),
      preStateHash: num32ToHex(r.pre_state_hash),
      postStateHash: num32ToHex(r.post_state_hash),
      timestamp: r.timestamp,
      computronsUsed: r.computrons_used,
      actionCount: r.action_count,
      previousReceiptHash: num32ToHex(r.previous_receipt_hash) ?? undefined,
      finality: typeof r.finality === "string" ? r.finality.toLowerCase() : undefined,
      wasEncrypted: r.was_encrypted,
      wasBurn: r.was_burn,
      raw: r,
      ...extra,
    });
  }

  /** The attached proof, if one has been attached (receipts are born proofless). */
  proof(): TurnProof | undefined {
    return this.attached;
  }

  /** Whether a proof has been attached. */
  hasProof(): boolean {
    return this.attached !== undefined;
  }

  /**
   * Attach the composed turn proof. Idempotent-at-first-writer: returns
   * `false` if one was already attached (a receipt never silently swaps
   * attestations) and throws [`WrongTurnProofError`] if the proof names a
   * different turn (a mis-bound attachment is refused, not stored).
   */
  attachProof(proof: TurnProof): boolean {
    const expected = hexDecodeExact(this.turnHash, 32);
    if (!bytesEqual(proof.turnHash, expected)) {
      throw new WrongTurnProofError(this.turnHash, proof.turnHashHex);
    }
    if (this.attached !== undefined) return false;
    this.attached = proof;
    return true;
  }

  /**
   * Lazily attach: return the attached proof, producing it with `f` if none
   * is attached yet (mirrors `Receipt::proof_or_attach`). A produced proof
   * bound to the wrong turn is refused, never stored.
   */
  async proofOrAttach(f: () => Promise<TurnProof>): Promise<TurnProof> {
    if (this.attached === undefined) {
      const produced = await f();
      this.attachProof(produced); // throws on wrong turn; false on race is fine
    }
    const got = this.attached;
    if (got === undefined) throw new Error("unreachable: attached above");
    return got;
  }
}

File: sdk/src/receipt.rs
//! # `Receipt` — the first of the SDK's two user-facing nouns.
//!
//! A [`Receipt`] is dregg's canonical proof-of-execution artifact: the
//! committed turn's hashes (turn / forest / effects), the pre/post state
//! roots, the agent cell, the federation binding, and the receipt-chain
//! link `previous_receipt_hash`. Every authorized submission
//! ([`AuthorizedTurn::submit`](crate::turns::AuthorizedTurn::submit))
//! returns one.
//!
//! The proof is **lazily attached**: a `Receipt` is born proofless (the
//! commit decision is the executor's; the STARK is additive attestation)
//! and a [`TurnProof`] — the composed full-turn STARK — can be attached
//! when a prover produces one (the node's async prove pool, a local
//! [`prove_full_turn`](crate::full_turn_proof::prove_full_turn) run, or a
//! fetched artifact). [`Receipt::proof()`] reads it; the internal proof
//! plumbing (`ComposedProof`, sub-proof components, PI layouts) stays off
//! the public surface.
//!
//! The second noun is [`AttestedHistory`](dregg_lightclient::AttestedHistory)
//! — the light-client artifact: the verdict obtained by verifying ONE
//! succinct whole-history aggregate. A `Receipt` says "this turn committed,
//! here is its place in a chain"; an `AttestedHistory` says "every turn of
//! this whole history executed correctly" — and between them that is the
//! entire user-facing proof story.

use std::sync::{Arc, OnceLock};

use dregg_turn::TurnReceipt;

use crate::full_turn_proof::{FullTurnProof, FullTurnVerifyError};

/// The proof attached to a [`Receipt`]: the composed full-turn STARK
/// (state transition + authorization + membership + conservation +
/// non-revocation in one verification), opaque on the public surface.
///
/// Obtain one from the proving pipeline
/// ([`prove_full_turn`](crate::full_turn_proof::prove_full_turn) /
/// [`prove_turn_self_sovereign`](crate::full_turn_proof::prove_turn_self_sovereign))
/// or by decoding a node-attached artifact, then attach it with
/// [`Receipt::attach_proof`].
#[derive(Clone, Debug)]
pub struct TurnProof(Arc<FullTurnProof>);

impl TurnProof {
    /// Wrap a composed full-turn proof.
    pub fn new(proof: FullTurnProof) -> Self {
        TurnProof(Arc::new(proof))
    }

    /// The turn hash this proof is bound to (replay protection: a proof
    /// only attests the turn it names).
    pub fn turn_hash(&self) -> [u8; 32] {
        self.0.turn_hash
    }

    /// The wire-serialized proof bytes.
    pub fn bytes(&self) -> &[u8] {
        &self.0.proof_bytes
    }

    /// Verify this proof against the expected pre/post state commitments.
    ///
    /// Thin delegation to
    /// [`verify_full_turn`](crate::full_turn_proof::verify_full_turn);
    /// freshness-critical verifiers (no-double-spend) should use
    /// [`verify_full_turn_bound`](crate::full_turn_proof::verify_full_turn_bound)
    /// on the inner proof via [`Self::inner`].
    pub fn verify(
        &self,
        expected_old_commit: [dregg_circuit::BabyBear; 8],
        expected_new_commit: [dregg_circuit::BabyBear; 8],
    ) -> Result<(), FullTurnVerifyError> {
        crate::full_turn_proof::verify_full_turn(&self.0, expected_old_commit, expected_new_commit)
    }

    /// Access the underlying composed proof (plumbing surface, for
    /// verifiers that need the bound entry points or component flags).
    pub fn inner(&self) -> &FullTurnProof {
        &self.0
    }
}

impl From<FullTurnProof> for TurnProof {
    fn from(p: FullTurnProof) -> Self {
        TurnProof::new(p)
    }
}

/// **The receipt noun.** A committed turn's proof-of-execution, with the
/// composed STARK proof lazily attached. See the module docs.
///
/// Dereferences to the wire-level [`TurnReceipt`], so every field and
/// method of the canonical receipt shape (`receipt_hash()`, `turn_hash`,
/// `post_state_hash`, …) is available directly on `Receipt`.
#[derive(Clone, Debug)]
pub struct Receipt {
    receipt: TurnReceipt,
    proof: OnceLock<TurnProof>,
}

impl Receipt {
    /// Wrap a committed receipt (proofless; attach later).
    pub fn new(receipt: TurnReceipt) -> Self {
        Receipt {
            receipt,
            proof: OnceLock::new(),
        }
    }

    /// The attached proof, if one has been attached (lazily — receipts are
    /// born proofless; the STARK is additive attestation produced by a
    /// prover after commit).
    pub fn proof(&self) -> Option<&TurnProof> {
        self.proof.get()
    }

    /// Whether a proof has been attached.
    pub fn has_proof(&self) -> bool {
        self.proof.get().is_some()
    }

    /// Attach the composed turn proof. Idempotent-at-first-writer: returns
    /// `Err` with the rejected proof if one was already attached (a receipt
    /// never silently swaps attestations) or if the proof names a different
    /// `turn_hash` than this receipt (a mis-bound attachment is refused,
    /// not stored).
    pub fn attach_proof(&self, proof: impl Into<TurnProof>) -> Result<(), TurnProof> {
        let proof = proof.into();
        if proof.turn_hash() != self.receipt.turn_hash {
            return Err(proof);
        }
        self.proof.set(proof)
    }

    /// Lazily attach: return the attached proof, producing it with `f` if
    /// none is attached yet. A produced proof bound to the wrong turn hash
    /// is refused (error), never stored.
    pub fn proof_or_attach<E>(
        &self,
        f: impl FnOnce() -> Result<TurnProof, E>,
    ) -> Result<&TurnProof, ProofAttachError<E>> {
        if self.proof.get().is_none() {
            let produced = f().map_err(ProofAttachError::Produce)?;
            if produced.turn_hash() != self.receipt.turn_hash {
                return Err(ProofAttachError::WrongTurn {
                    expected: self.receipt.turn_hash,
                    got: produced.turn_hash(),
                });
            }
            // A racing attach of the same receipt's proof is fine; first
            // writer wins and we read whatever landed.
            let _ = self.proof.set(produced);
        }
        Ok(self.proof.get().expect("set above or already present"))
    }

    /// The wire-level receipt (canonical serialization shape).
    pub fn as_turn_receipt(&self) -> &TurnReceipt {
        &self.receipt
    }

    /// Unwrap into the wire-level receipt (drops any attached proof).
    pub fn into_turn_receipt(self) -> TurnReceipt {
        self.receipt
    }
}

/// Why [`Receipt::proof_or_attach`] failed.
#[derive(Debug)]
pub enum ProofAttachError<E> {
    /// The producer closure failed.
    Produce(E),
    /// The produced proof is bound to a different turn than this receipt.
    WrongTurn { expected: [u8; 32], got: [u8; 32] },
}

impl From<TurnReceipt> for Receipt {
    fn from(receipt: TurnReceipt) -> Self {
        Receipt::new(receipt)
    }
}

impl std::ops::Deref for Receipt {
    type Target = TurnReceipt;
    fn deref(&self) -> &TurnReceipt {
        &self.receipt
    }
}

impl serde::Serialize for Receipt {
    /// Serializes as the wire-level [`TurnReceipt`] — the proof travels
    /// separately (it is attestation about the receipt, not part of it).
    fn serialize<S: serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
        self.receipt.serialize(s)
    }
}

impl<'de> serde::Deserialize<'de> for Receipt {
    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
        Ok(Receipt::new(TurnReceipt::deserialize(d)?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn receipt_derefs_and_roundtrips() {
        let mut tr = TurnReceipt::default();
        tr.turn_hash = [7u8; 32];
        let r = Receipt::new(tr.clone());
        assert_eq!(r.turn_hash, [7u8; 32]);
        assert_eq!(r.receipt_hash(), tr.receipt_hash());
        assert!(!r.has_proof());
        let json = serde_json::to_string(&r).unwrap();
        let back: Receipt = serde_json::from_str(&json).unwrap();
        assert_eq!(back.turn_hash, [7u8; 32]);
    }
}
